### PR TITLE
Feat/add nlb

### LINKED ISSUE #66: Add option for network type loadbalancer.
Add option for network type loadbalancer.

https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/environments-cfg-nlb.html

### PR DESCRIPTION
**Pull request by florian0410**
## what
* Add NLB support in the module
* Set default protocol to TCP in case that loadbalancer_type == "network"
* S3 logs and Security Groups are not valid for Network ELB. 
* `HealthCheckPath` apply to to ALB only.

## why
* Module does not support yet NLB integration

## references
* closes #66 
* closes #176
* closes #105   
* closes #101

### CODE REVIEW

**Comment by florian0410** [comment]
I think we need another example case for nlb.

The StickinessEnabled additional setting of alb is not compatible with NLB currently.

Does the CI handle a multi example case ? I don't think I can set it up alone here.

**Comment by korenyoni** [comment]
> I think we need another example case for nlb.
> 
> The StickinessEnabled additional setting of alb is not compatible with NLB currently.
> 
> Does the CI handle a multi example case ? I don't think I can set it up alone here.

@florian0410 Yes.

Let's say you are going to add a new example directory `examples/nlb`, you should also add a `examples_nlb_test.go` file to `test/src/` and also invoke it right below https://github.com/cloudposse/terraform-aws-elastic-beanstalk-environment/blob/master/test/src/Makefile#L17

> I also need help for the bridgecrew error.

Without digging too deep, this seems like a BridgeCrew bug because `s3_bucket_versioning_enabled` is set to true by default. I would add `#bridgecrew:skip=BC_AWS_S3_16` to the `aws_s3_bucket.elb_logs` resource block and reference this PR in the Bridgecrew policy skip description. Better yet, I would open up an issue in https://github.com/bridgecrewio/checkov and reference the issue in the Bridgecrew policy skip description.

**Comment by florian0410** [comment]
I just added a new example following your advices. I'm not fully used to go yet so I may have done errors but I should have added the essential.

For the bridgecrew I added the exception, but I didn't open an issue for now since I'm not used to the tool. I'll let you handle this one if you don't mind. After a quick search since I use a count condition I think it may be linked to the following issue: https://github.com/bridgecrewio/checkov/issues/1078

But not sure yet so I'll let you check this.

@korenyoni

**Comment by korenyoni** [comment]
@florian0410 see: https://github.com/cloudposse/actions/runs/2831747735?check_suite_focus=true#step:9:2094

**Comment by florian0410** [comment]
Fixed @korenyoni I rebased also to the new version 0.41.0 by the way.

File: examples/nlb/fixtures.us-east-2.tfvars [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+force_destroy = true
+
+rolling_update_enabled = true
+
+rolling_update_type = "Health"
+
+updating_min_in_service = 0
+
+updating_max_batch = 1
+
```
Comment (korenyoni) [inline examples/nlb/fixtures.us-east-2.tfvars:39]:
```suggestion
updating_max_batch = 1

healthcheck_url = "/"

```

File: examples/nlb/main.tf [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  autoscale_unit            = var.autoscale_unit
+  autoscale_lower_bound     = var.autoscale_lower_bound
+  autoscale_lower_increment = var.autoscale_lower_increment
+  autoscale_upper_bound     = var.autoscale_upper_bound
+  autoscale_upper_increment = var.autoscale_upper_increment
+
+  vpc_id                  = module.vpc.vpc_id
+  loadbalancer_subnets    = module.subnets.public_subnet_ids
+  application_subnets     = module.subnets.private_subnet_ids
+  allowed_security_groups = [module.vpc.vpc_default_security_group_id]
```
Comment (korenyoni) [inline examples/nlb/main.tf:67]:
```suggestion
  security_group_rules = [
    {
      type                     = "egress"
      from_port                = 0
      to_port                  = 65535
      protocol                 = "-1"
      cidr_blocks              = ["0.0.0.0/0"]
      source_security_group_id = null
      description              = "Allow all outbound traffic"
    },
    {
      type                     = "ingress"
      from_port                = 0
      to_port                  = 65535
      protocol                 = "-1"
      cidr_blocks              = []
      source_security_group_id = module.vpc.vpc_default_security_group_id
      description              = "Allow all inbound traffic from trusted Security Groups"
    },
  ]
```

File: examples/nlb/fixtures.us-east-2.tfvars [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+availability_zone_selector = "Any 2"
+
+instance_type = "t3.micro"
+
+autoscale_min = 1
+
+autoscale_max = 2
+
+wait_for_ready_timeout = "20m"
+
```
Comment (korenyoni) [inline examples/nlb/fixtures.us-east-2.tfvars:28]:
```suggestion
```

**Review by korenyoni — CHANGES_REQUESTED** [review]
Still failing https://github.com/cloudposse/actions/runs/2832178965?check_suite_focus=true#step:9:2008

File: examples/nlb/fixtures.us-east-2.tfvars (hunk 1 above)
Comment (florian0410) [reply examples/nlb/fixtures.us-east-2.tfvars:39]: healthcheck_url  is not compatible with NLB so I would remove it instead.

**Comment by florian0410** [comment]
Any need or remarks for the review ?
@korenyoni

**Comment by florian0410** [comment]
Added one of your suggestion @korenyoni.

Checks should succeed, normally. Do you know if we could merge this one soon ?

**Comment by Gowiem** [comment]
@florian0410 you're one of the primary contributors to this module... would you be interested in becoming a maintainer for it? I believe @osterman is still looking around for someone.

**Comment by florian0410** [comment]
> @florian0410 you're one of the primary contributors to this module... would you be interested in becoming a maintainer for it? I believe @osterman is still looking around for someone.

I think I can give it a try yes. You can ping me on slack to talk about it

**Comment by osterman** [comment]
@florian0410 looks like we have too many "florians" on SweetOps, so I cannot find you. Can you DM me on slack?

**Comment by florian0410** [comment]
Hello,

Can we run again the tests ? It sounds like previous checks encountered a random NAT issue and I added some enhancement to this PR since last time.
I cannot see the bridgecrew  error also.


Regards

**Review by Gowiem — APPROVED** [review]
Looks solid to me @florian0410 💯 

If you want a more keen set of eyes on this work before merging then you can wait for a final review from @korenyoni as it seems he's been on top of this one so far.

**Comment by florian0410** [comment]
Just released version 0.42.0.

Thank you all for your help and reviews !

**Pull request merged**